controllers/network/ipset_controller.go

```diff
@@ -0,0 +1,446 @@
+			return
+		}
+	}()
+
+	// If we're not deleting this and the service object doesn't have our finalizer, add it.
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
+		return ctrl.Result{}, err
+	}
+
+	// initialize status
+	if instance.Status.Conditions == nil {
+		instance.Status.Conditions = condition.Conditions{}
+
+		cl := condition.CreateList(
+			condition.UnknownCondition(condition.InputReadyCondition, condition.InitReason, condition.InputReadyInitMessage),
+			condition.UnknownCondition(networkv1.ReservationReadyCondition, condition.InitReason, networkv1.ReservationInitMessage),
+		)
+
+		instance.Status.Conditions.Init(&cl)
+
+		// Register overall status immediately to have an early feedback e.g. in the cli
+		return ctrl.Result{}, nil
+	}
+
+	instance.Status.Reservation = []networkv1.IPAddress{}
+
+	// Handle service delete
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper)
+	}
+
+	// Handle non-deleted ipsets
+	return r.reconcileNormal(ctx, instance, helper)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *IPSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	ipsetFN := handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+		result := []reconcile.Request{}
+
+		// For each NetConfig update event get the list of all
+		// IPSet to trigger reconcile for the one in the same namespace
+		ipsets := &networkv1.IPSetList{}
+
+		listOpts := []client.ListOption{
+			client.InNamespace(o.GetNamespace()),
+		}
+		if err := r.Client.List(context.Background(), ipsets, listOpts...); err != nil {
+			r.Log.Error(err, "Unable to retrieve IPSetList %w")
+			return nil
+		}
+
+		// For each ipsets instance create a reconcile request
+		for _, i := range ipsets.Items {
+			name := client.ObjectKey{
+				Namespace: o.GetNamespace(),
+				Name:      i.Name,
+			}
+			result = append(result, reconcile.Request{NamespacedName: name})
+		}
+		if len(result) > 0 {
+			r.Log.Info(fmt.Sprintf("Reconcile request for: %+v", result))
+
+			return result
+		}
+		return nil
+	})
+
+	// 'UpdateFunc' used to judge if a event about the object is what we want.
+	// If that is true, the event will be processed by the reconciler.
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.ObjectOld.GetResourceVersion() != e.ObjectNew.GetResourceVersion()
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&networkv1.IPSet{}).
+		Owns(&networkv1.Reservation{}).
+		Watches(&source.Kind{Type: &networkv1.NetConfig{}},
+			ipsetFN,
+			builder.WithPredicates(p)).
+		Complete(r)
+}
+
+func (r *IPSetReconciler) reconcileDelete(ctx context.Context, instance *networkv1.IPSet, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service delete")
+
+	// get reservation
+	res, err := r.getReservation(ctx, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
I think it's possible to get a `nil` `res` here given L320-323 (but `nil` `err` as well, so it won't return on L208).  Maybe we should check for that before calling `controllerutil.RemoveFinalizer()`?

Reply:
ack, will address that. probably change getReservation and handle NotFound in the upper level

Reply:
done